File: internal/schema/schema_test.go
package schema

// Copyright (C) 2023 by Posit Software, PBC.
import (
	"fmt"
	"strings"
	"testing"

	"github.com/spf13/afero"
	"github.com/stretchr/testify/suite"

	"github.com/posit-dev/publisher/internal/types"
	"github.com/posit-dev/publisher/internal/util"
	"github.com/posit-dev/publisher/internal/util/utiltest"
)

type SchemaSuite struct {
	utiltest.Suite
	cwd util.AbsolutePath
}

func TestSchemaSuite(t *testing.T) {
	suite.Run(t, new(SchemaSuite))
}

func (s *SchemaSuite) SetupTest() {
	cwd, err := util.Getwd(nil)
	s.NoError(err)
	s.cwd = cwd
}

type genericContent map[string]any

type fileValidationTestCase struct {
	schemaURL string
	dataFile  []string
}

func (s *SchemaSuite) TestValidateFile() {
	cases := []fileValidationTestCase{
		{
			schemaURL: ConfigSchemaURL,
			dataFile:  []string{"config.toml"},
		},
		{
			schemaURL: ConfigSchemaURL,
			dataFile:  []string{"config_cloud.toml"},
		},
		{
			schemaURL: "https://cdn.posit.co/publisher/schemas/draft/posit-publishing-schema-v3.json",
			dataFile:  []string{"draft", "config_cloud.toml"},
		},
		{
			schemaURL: DeploymentSchemaURL,
			dataFile:  []string{"record.toml"},
		},
		{
			schemaURL: DeploymentSchemaURL,
			dataFile:  []string{"record_cloud.toml"},
		},
		{
			schemaURL: "https://cdn.posit.co/publisher/schemas/draft/posit-publishing-record-schema-v3.json",
			dataFile:  []string{"draft", "record.toml"},
		},
	}

	for _, testCase := range cases {
		s.Run(strings.Join(testCase.dataFile, "/"), func() {
			validator, err := NewValidator[genericContent](testCase.schemaURL)
			s.NoError(err)
			schemaPath := append([]string{"schemas"}, testCase.dataFile...)
			path := s.cwd.Join(schemaPath...)
			err = validator.ValidateTOMLFile(path)
			s.NoError(err)
		})
	}
}

type dataValidationTestCase struct {
	title string
	data  map[string]any
}

func (s *SchemaSuite) TestValidateConfig() {
	cases := []dataValidationTestCase{
		{
			title: "connect config without product_type",
			data: map[string]any{
				"$schema":    ConfigSchemaURL,
				"type":       "python-dash",
				"entrypoint": "app.py",
				"python": map[string]any{
					"requires_python": ">=3.8",
				},
				"connect": map[string]any{},
			},
		},
		{
			title: "connect config with product_type",
			data: map[string]any{
				"$schema":      ConfigSchemaURL,
				"product_type": "connect",
				"type":         "python-dash",
				"entrypoint":   "app.py",
				"python": map[string]any{
					"requires_python": ">=3.8",
				},
				"connect": map[string]any{},
			},
		},
		{
			title: "connect cloud config with product_type",
			data: map[string]any{
				"$schema":      ConfigSchemaURL,
				"product_type": "connect_cloud",
				"type":         "python-dash",
				"entrypoint":   "app.py",
				"python": map[string]any{
					"version": "3.8",
				},
				"connect_cloud": map[string]any{},
			},
		},
	}

	for _, testCase := range cases {
		s.Run(testCase.title, func() {
			validator, err := NewValidator[genericContent](ConfigSchemaURL)
			s.NoError(err)
			err = validator.ValidateContent(testCase.data)
			s.NoError(err)
		})
	}
}

func (s *SchemaSuite) TestValidationError() {
	cwd, err := util.Getwd(afero.NewMemMapFs())
	s.NoError(err)
	err = cwd.MkdirAll(0700)
	s.NoError(err)

	badTOML := []byte("bad-attr = 1\n")
	path := cwd.Join("test.toml")
	err = path.WriteFile(badTOML, 0600)
	s.NoError(err)

	validator, err := NewValidator[genericContent](ConfigSchemaURL)
	s.NoError(err)
	err = validator.ValidateTOMLFile(path)
	agentErr, ok := err.(*types.AgentError)
	s.True(ok)
	s.Equal(agentErr.Code, tomlValidationErrorCode)
}

type invalidPropertyTestCase struct {
	productType string
	basePath    []string
	propName    string
	propValue   any

	message string
}

func (s *SchemaSuite) TestDisallowedProperties() {
	cases := []invalidPropertyTestCase{
		// Root level invalid properties
		{
			productType: "connect",
			basePath:    []string{},
			propName:    "garbage",
			propValue:   "value",
			message:     "garbage: not allowed.",
		},

		// Connect product_type invalid properties
		{
			productType: "connect",
			basePath:    []string{"python"},
			propName:    "garbage",
			propValue:   "value",
			message:     "python: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"r"},
			propName:    "garbage",
			propValue:   "value",
			message:     "r: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"jupyter"},
			propName:    "garbage",
			propValue:   "value",
			message:     "jupyter: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"quarto"},
			propName:    "garbage",
			propValue:   "value",
			message:     "quarto: missing properties: 'version'; quarto: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"connect"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"connect", "runtime"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect.runtime: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"connect", "kubernetes"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect.kubernetes: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect",
			basePath:    []string{"connect", "access"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect.access: additionalProperties 'garbage' not allowed.",
		},

		// Connect Cloud product_type invalid properties
		{
			productType: "connect_cloud",
			basePath:    []string{"python"},
			propName:    "garbage",
			propValue:   "value",
			message:     "python: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"python"},
			propName:    "requires_python",
			propValue:   ">=3.8",
			message:     "python: additionalProperties 'requires_python' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"python"},
			propName:    "package_file",
			propValue:   "requirements.txt",
			message:     "python: additionalProperties 'package_file' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"python"},
			propName:    "package_manager",
			propValue:   "pip",
			message:     "python: additionalProperties 'package_manager' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"r"},
			propName:    "garbage",
			propValue:   "value",
			message:     "r: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"r"},
			propName:    "requires_r",
			propValue:   ">=4.2",
			message:     "r: additionalProperties 'requires_r' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"r"},
			propName:    "package_file",
			propValue:   "renv.lock",
			message:     "r: additionalProperties 'package_file' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"r"},
			propName:    "package_manager",
			propValue:   "renv",
			message:     "r: additionalProperties 'package_manager' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"connect_cloud"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect_cloud: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"connect_cloud", "python"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect_cloud.python: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"connect_cloud", "r"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect_cloud.r: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{"connect_cloud", "access_control"},
			propName:    "garbage",
			propValue:   "value",
			message:     "connect_cloud.access_control: additionalProperties 'garbage' not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{},
			propName:    "jupyter",
			propValue:   map[string]any{},
			message:     "jupyter: not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{},
			propName:    "quarto",
			propValue:   map[string]any{},
			message:     "quarto: not allowed.",
		},
		{
			productType: "connect_cloud",
			basePath:    []string{},
			propName:    "has_parameters",
			propValue:   true,
			message:     "has_parameters: not allowed.",
		},

		// Connect properties shouldn't be allowed with connect_cloud product_type
		{
			productType: "connect_cloud",
			basePath:    []string{},
			propName:    "connect",
			propValue:   map[string]any{},
			message:     "connect: not allowed.",
		},

		// Connect_cloud properties shouldn't be allowed with connect product_type
		{
			productType: "connect",
			basePath:    []string{},
			propName:    "connect_cloud",
			propValue:   map[string]any{},
			message:     "connect_cloud: not allowed.",
		},
	}

	for _, tc := range cases {
		propPath := append(tc.basePath, tc.propName)
		description := fmt.Sprintf("%s:  .%s=%s", tc.productType, strings.Join(propPath, "."), tc.propValue)
		s.Run(description, func() {
			// Set up the validator
			validator, err := NewValidator[genericContent](ConfigSchemaURL)
			s.NoError(err)

			// Create a valid base configuration map
			baseConfig := map[string]any{
				"$schema":      ConfigSchemaURL,
				"product_type": tc.productType,
				"type":         "html",
				"entrypoint":   "index.html",
			}

			// Ensure it's valid
			err = validator.ValidateContent(baseConfig)
			s.NoError(err, "Base config should be valid before adding invalid property")

			// Acquire the nested structure
			target := baseConfig
			for _, path := range tc.basePath {
				if _, exists := target[path]; !exists {
					target[path] = map[string]any{}
				}
				target = target[path].(map[string]any)
			}

			// Add the invalid property
			target[tc.propName] = tc.propValue

			// Validate and expect an error
			err = validator.ValidateContent(baseConfig)
			s.Error(err, "Expected validation error")

			// Check that it's a schema validation error
			agentErr, ok := err.(*types.AgentError)
			s.True(ok)
			s.Equal(tomlValidationErrorCode, agentErr.Code)
			s.Equal(tc.message, agentErr.Message)
		})
	}
}

File: extensions/vscode/src/toml/schema.test.ts
// Copyright (C) 2026 by Posit Software, PBC.

import fs from "fs";
import path from "path";

import { describe, expect, it } from "vitest";
import Ajv2020 from "ajv/dist/2020";
import addFormats from "ajv-formats";
import { parse as parseTOML } from "smol-toml";

import schema from "./schemas/posit-publishing-schema-v3.json";
import recordSchema from "./schemas/posit-publishing-record-schema-v3.json";
import draftSchema from "./schemas/draft/posit-publishing-schema-v3.json";
import draftRecordSchema from "./schemas/draft/posit-publishing-record-schema-v3.json";

const ajv = new Ajv2020({ strict: false, allErrors: true });
addFormats(ajv);
const validate = ajv.compile(schema);

// Helper: create a minimal valid config for a given product_type
function baseConfig(productType: string): Record<string, unknown> {
  return {
    $schema:
      "https://cdn.posit.co/publisher/schemas/posit-publishing-schema-v3.json",
    product_type: productType,
    type: "html",
    entrypoint: "index.html",
  };
}

// Helper: set a nested property on an object, creating intermediate objects as needed
function setNested(
  obj: Record<string, unknown>,
  path: string[],
  key: string,
  value: unknown,
): void {
  let target = obj;
  for (const segment of path) {
    if (target[segment] === undefined) {
      target[segment] = {};
    }
    target = target[segment] as Record<string, unknown>;
  }
  target[key] = value;
}

function expectValid(data: Record<string, unknown>): void {
  const valid = validate(data);
  expect(
    valid,
    `Expected valid but got errors: ${JSON.stringify(validate.errors)}`,
  ).toBe(true);
}

function expectInvalid(data: Record<string, unknown>): void {
  const valid = validate(data);
  expect(valid, "Expected validation to fail").toBe(false);
}

describe("schema validates valid configs", () => {
  it("connect config without product_type", () => {
    const data = baseConfig("connect");
    delete data.product_type;
    expectValid(data);
  });

  it("connect config with product_type", () => {
    expectValid(baseConfig("connect"));
  });

  it("connect cloud config with product_type", () => {
    const data = baseConfig("connect_cloud");
    data.python = { version: "3.11" };
    expectValid(data);
  });

  it("connect config with integration_requests", () => {
    const data = baseConfig("connect");
    data.integration_requests = [
      { guid: "12345678-1234-1234-1234-1234567890ab" },
    ];
    expectValid(data);
  });
});

describe("schema validates Node.js configs", () => {
  it("accepts type=nodejs without a [node] section", () => {
    const data = baseConfig("connect");
    data.type = "nodejs";
    data.entrypoint = "index.js";
    expectValid(data);
  });

  it("rejects [node] section under any product_type", () => {
    const data = baseConfig("connect");
    data.type = "nodejs";
    data.entrypoint = "index.js";
    data.node = {};
    expectInvalid(data);
  });
});

describe("schema rejects unknown property at root", () => {
  it("rejects unknown root property", () => {
    const data = baseConfig("connect");
    data.garbage = "value";
    expectInvalid(data);
  });
});

describe("schema rejects invalid integration_requests", () => {
  it("rejects integration_requests as string instead of array", () => {
    const data = baseConfig("connect");
    data.integration_requests = "string-instead-of-array";
    expectInvalid(data);
  });

  it("rejects integration_requests with string instead of object", () => {
    const data = baseConfig("connect");
    data.integration_requests = ["string-instead-of-object"];
    expectInvalid(data);
  });
});

describe("schema rejects disallowed properties for Connect", () => {
  it("rejects python.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["python"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects r.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["r"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects jupyter.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["jupyter"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects quarto.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["quarto"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["connect"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect.runtime.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["connect", "runtime"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect.kubernetes.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["connect", "kubernetes"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect.access.garbage", () => {
    const data = baseConfig("connect");
    setNested(data, ["connect", "access"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect_cloud section with connect product_type", () => {
    const data = baseConfig("connect");
    data.connect_cloud = {};
    expectInvalid(data);
  });
});

describe("schema rejects disallowed properties for Connect Cloud", () => {
  it("rejects python.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["python"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects python.requires_python", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["python"], "requires_python", ">=3.8");
    expectInvalid(data);
  });

  it("rejects python.package_file", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["python"], "package_file", "requirements.txt");
    expectInvalid(data);
  });

  it("rejects python.package_manager", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["python"], "package_manager", "pip");
    expectInvalid(data);
  });

  it("rejects r.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["r"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects r.requires_r", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["r"], "requires_r", ">=4.2");
    expectInvalid(data);
  });

  it("rejects r.package_file", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["r"], "package_file", "renv.lock");
    expectInvalid(data);
  });

  it("rejects r.package_manager", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["r"], "package_manager", "renv");
    expectInvalid(data);
  });

  it("rejects connect_cloud.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["connect_cloud"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect_cloud.python.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["connect_cloud", "python"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect_cloud.r.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["connect_cloud", "r"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects connect_cloud.access_control.garbage", () => {
    const data = baseConfig("connect_cloud");
    setNested(data, ["connect_cloud", "access_control"], "garbage", "value");
    expectInvalid(data);
  });

  it("rejects jupyter section", () => {
    const data = baseConfig("connect_cloud");
    data.jupyter = {};
    expectInvalid(data);
  });

  it("rejects quarto section", () => {
    const data = baseConfig("connect_cloud");
    data.quarto = {};
    expectInvalid(data);
  });

  it("rejects has_parameters", () => {
    const data = baseConfig("connect_cloud");
    data.has_parameters = true;
    expectInvalid(data);
  });

  it("rejects connect.kubernetes", () => {
    const data = baseConfig("connect_cloud");
    data.connect = { kubernetes: {} };
    expectInvalid(data);
  });
});

describe("example TOML files validate against their schemas", () => {
  const schemasDir = path.resolve(__dirname, "schemas");

  function loadTOML(relativePath: string): Record<string, unknown> {
    const content = fs.readFileSync(
      path.join(schemasDir, relativePath),
      "utf-8",
    );
    return parseTOML(content) as Record<string, unknown>;
  }

  function compileWithRef(
    mainSchema: Record<string, unknown>,
    refSchema: Record<string, unknown>,
  ) {
    const localAjv = new Ajv2020({ strict: false, allErrors: true });
    addFormats(localAjv);
    localAjv.addSchema(refSchema);
    return localAjv.compile(mainSchema);
  }

  // Release config schema
  const validateConfig = validate;
  // Release record schema (references config schema via $ref)
  const validateRecord = compileWithRef(recordSchema, schema);
  // Draft config schema
  const validateDraftConfig = (() => {
    const draftAjv = new Ajv2020({ strict: false, allErrors: true });
    addFormats(draftAjv);
    return draftAjv.compile(draftSchema);
  })();
  // Draft record schema (references draft config schema via $ref)
  const validateDraftRecord = compileWithRef(draftRecordSchema, draftSchema);

  const cases: { file: string; validator: ReturnType<typeof ajv.compile> }[] = [
    { file: "example-config.toml", validator: validateConfig },
    { file: "example-config-cloud.toml", validator: validateConfig },
    { file: "example-record.toml", validator: validateRecord },
    { file: "example-record-cloud.toml", validator: validateRecord },
    {
      file: "draft/example-config.toml",
      validator: validateDraftConfig,
    },
    {
      file: "draft/example-config-cloud.toml",
      validator: validateDraftConfig,
    },
    { file: "draft/example-record.toml", validator: validateDraftRecord },
  ];

  for (const { file, validator } of cases) {
    it(file, () => {
      const data = loadTOML(file);
      const valid = validator(data);
      expect(
        valid,
        `${file} failed validation: ${JSON.stringify(validator.errors, null, 2)}`,
      ).toBe(true);
    });
  }
});
